### PR TITLE
Add script to swap ERC-20 token balances in Gitcoin Fee address to ETH using Uniswap

### LINKED ISSUE #4150: As a Gitcoin administrator, I would like any ERC20 tokens in my FEE_ADDRESS converted to ETH so I don't have to manage a complex tax situation
### User Story

[comment]: # (As a <user type>, I want to <task> so that <goal>.)
As a Gitcoin administrator, I would like any ERC20 tokens in my FEE_ADDRESS converted to ETH so I don't have to manage a complex tax situation

### Why Is this Needed
because john chard advised us taxes can be a pain in the arse.

[comment]: # (Describe the problem and why this task is needed. Provide description of the current state, what you would like to happen, and what actually happen)
*Summary*:

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*:
feature

### Current Behavior
[comment]: # (Describe what actually happened.)
no management of ERC20 tokens in FEE_ADDRESS

### Expected Behavior
[comment]: # (Describe what you expected to happen.)
a new django management command that is scheduled to run every hour.
two new entries into `.env` and `settings.py` which manage the following vars
- FEE_ADDRESS -- the web3 address where fees exist
- FEE_ADDRESS_PRIVATE_KEY -- the private key to the above address.

the management command will, when run, check for any ERC20 tokens and use uniswap to transfer them into ETH.

### Definition of Done
[comment]: # (Any other information that would be useful, bullets are helpful.)
see above ^^

### Data Requirements
[comment]: # (How will we measure the success of this feature? What kind of tracking is needed for this feature (clicks, impressions, flag)?)

### Additional Information
[comment]: # (Any other information that would be useful, content, screenshots, etc.)

### PR DESCRIPTION
**Pull request by acolytec3**
##### Description

Pulls ERC-20 token balances associated with fee address and converts to ETH using Uniswap.
Notes:
 * On Rinkeby testnet, Uniswap trading request only currently works for BAT Uniswap exchange but not OMG or MKR exchanges (using my on test account) so would appreciate any comment or insight on this.
 * Needs comment on what is desired in terms of tracking token balances that don't have corresponding Uniswap exchange.  Currently just ignores them.

##### Checklist

- [ ] Read and conforms to the [contributor guidelines](https://docs.gitcoin.co/mk_contributors/)
- [ ] Read and conforms to the [style guidelines](https://docs.gitcoin.co/mk_styleguide/)
- [ ] Linter status: 100% pass
- [ ] Added relevant tests
- [ ] Tests pass and code coverage has not decreased
- [ ] Changes don't break existing behavior
- [ ] Commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)
- [ ] [Code documentation](https://docs.gitcoin.co/mk_contributors/#docstrings)
- [ ] Completed: Code review by core team and any requested changes

##### Affected core subsystem(s)

N/A

##### Refers/Fixes

Fixes #4150 

###### Contributor

- [x ] Read and followed the [Contributor Guidelines](https://docs.gitcoin.co/mk_contributors/)
- [x ] Tested all changes **locally**
- [x ] Verified existing functionality
- [x] Ran `make test` and everything passed!

###### Reviewer

- [ ] Affirm contributor guidelines have been followed and requested changes made
- [ ] CI tests and linting pass
- [ ] No conflicts (migrations, files, etc)
- [ ] Regression tested against staging or local deployment

###### Funder

- [ ] Validated requested changes were made to specification
- [ ] Bounty payout released to the contributor

### CODE REVIEW

**Comment by acolytec3** [comment]
Still having trouble with the Uniswap trade function.  Any insights?

**Comment by acolytec3** [comment]
One follow up about the uniswap trading issue, using an address of my own on the Rinkeby network, this script will now successfully swap BAT balances to ETH but the same transaction fails for Uniswap exchange contracts for OMG and MKR balances.  So far, I haven't been able to detect any reason why it works for one ERC-20 token and not others and haven't seen anything else online where people have seen similar issues.  Would appreciate any insights.

**Comment by acolytec3** [comment]
Also, all my tests pass locally using the "pytest -p no:ethereum" test command in the docker container per the below screenshot.  Not sure why Travis is failing it.
image

**Comment by owocki** [comment]
code looks ssane to me.  would love to see feedback from the other engineers tho

can we schedule this to run once per day in the `crontab` file?

**Comment by acolytec3** [comment]
A few updates:

1) Added a schedule in the crontab to run every hour.
2) I fixed a logic issue in constructing the web3 provider URL that wasn't obvious to me before
3) I removed the need for the test flag.  The script now determines which network to use based on the WEB3_HTTP_PROVIDER field and sets necessary parameters internally as necessary.

I'm still seeing the uniswap sell token method failing for the OMG and MKR tokens in my wallet on rinkeby but I'm at a loss as to why this should be the case.  Given that this is kind of the point of the whole exercise. would appreciate any insights why it's working for BAT and not OMG/MKR.

**Comment by owocki** [comment]
ok cool; i think we'll want to deploy this but with a test account so we can be sure it's working. 

one thing i forgot to add to the original spec was writing some persistent record of the token transfer to the DB.  maybe call the table `CurrencyConversion` and have it have the following fields
- created_on
- from_amount
- to_amount
- from_token_addr
- to_token_addr
- from_token_symbol
- to_token_symbol

etc

i know this was not in the original scope.  maybe i could add a nice tip on top of the existing amount to make it worth your while @acolytec3 ?

**Comment by owocki** [comment]
also @danlipert @thelostone-mc if u have any extra time this sprint i could use a review of this code

**Comment by acolytec3** [comment]
@owocki Let me see what I can do.  Is it something you'd want as a new model under the revenue app or somewhere else?

**Comment by owocki** [comment]
revenue app is fine

**Comment by acolytec3** [comment]
@owocki I've added the new CurrencyConversion table and also the SendGrid notification if a transaction fails as requested.

One other thing I was thinking about along these lines, if a transaction fails, would it make sense to still insert a row into the table with a transaction status of failed?  I could then add some additional logic into the sell_token method to check and see if a previous transaction for that token failed so it wouldn't try it again and burn gas pointlessly since this job runs hourly.  Since I don't know if the transaction failures I'm seeing for OMG/MKR/DAI on Rinkeby are some unidentified issue with my code or with the Uniswap exchange contracts on Rinkeby, I don't know if this is a real failure scenario or not.

**Comment by owocki** [comment]
> One other thing I was thinking about along these lines, if a transaction fails, would it make sense to still insert a row into the table with a transaction status of failed?

yes i think that makes sense.

**Comment by acolytec3** [comment]
@owocki Okay, added an additional field to the CurrencyConversion table to note transaction success/failure and an additional check in the sell_token function to check the database for a failed transaction record prior to attempting to convert the token to ETH.  If it finds one, it sends a an email saying that it found a previous failed transaction so not attempting to convert again.  

So, this will at least keep the Fee wallet from burning gas pointlessly (assuming that we see similar issues with some uniswap exchanges on mainnet).  But, it does bring a couple of further items to think through:
1) The way it's currently designed, this will send an email on every failed transaction or where the script sees a failed transaction in the CurrencyConversion table.  This means that someone will have to go and delete that failed transaction from the database in order for the script to try and convert that token to ETH again.  It also means you'll get one or more emails every hour nagging you about failed transactions.  Not sure if this is desirable behavior or not.

2) Do you want me to hardcode the "to" address for the transaction failure emails?  Right now, I have it set to use the SERVER_EMAIL environmental variable but can easily hard code founders@gitcoin.co or else create a new variable for it as needed.  I just didn't want every other dev who happens to run a local version of this to be accidentally spamming that address if someone forgets to change it in their code.

Otherwise, I think this is close to done.  I'm going to work on the test coverage a little more but I think all the functionality is done and working (at least using the wallet I tested with on Rinkeby).

**Comment by owocki** [comment]
1. thats perefect; thank you!
2. `settings.CONTACT_EMAIL` is fine 

appreciate the thoughtful questions and attention to detail on this here

**Comment by acolytec3** [comment]
Will do what I can.  My attempt to rebase this morning went awry somewhere so trying to figure out what git didn't like.

**Comment by acolytec3** [comment]
I did finally get the rebase to work and then validated that it's still working as expected.  My attempt at squashing doesn't seem to have worked though.  Is that something you can do once you decide it's ready to merge?  If not, I can try again, but something funny is happening in my commit history whenever I try to merge.

**Comment by thelostone-mc** [comment]
@acolytec3 Ah the commit history in the PR is messed and makes reviewing really hard ! 
Could you cherry-pick the commits + squash + and we could review it ?

cc @gitcoinco/engineers

**Comment by acolytec3** [comment]
Understood. Let me try again and see if I can clean it up. I think rebasing
threw everything off.

**Comment by acolytec3** [comment]
I'm going to close this PR and start a fresh one that will hopefully only have my code.  I can't for the life of me figure out how to clean the mess introduced when I rebased.
